### ISSUE TITLE
Vault Leader doesn't step down if it can't communicate with DynamoDB backend 

### ISSUE THREAD

**Issue by nemosupremo** [hashicorp/vault#5045]
**Describe the bug**
I've had a HA Vault setup running (2 nodes) for around 10 days. Suddenly, unexpectedly, every request (token requests, token refreshes, even `/sys/leader`) was failing for several minutes with:

```
* NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

I first thought it could be related to the AWS User role I have for vault for the `/aws` secret backend, I rotated those keys, but no luck. I then restarted both Vault nodes, unsealed one, and operation continued as normal. I then noticed I couldn't restart one of the Vault nodes:

```
Aug 03 22:52:09 ip-172-31-21-132 systemd[1]: Stopped Vault service.
Aug 03 22:52:09 ip-172-31-21-132 systemd[1]: Starting Vault service...
Aug 03 22:52:09 ip-172-31-21-132 systemd[1]: Started Vault service.
Aug 03 22:52:09 ip-172-31-21-132 vault[27133]: Error initializing storage of type dynamodb: NoCredentialProviders: no valid providers in chain. Deprecated.
Aug 03 22:52:09 ip-172-31-21-132 vault[27133]:         For verbose messaging see aws.Config.CredentialsChainVerboseErrors
Aug 03 22:52:09 ip-172-31-21-132 systemd[1]: vault.service: Main process exited, code=exited, status=1/FAILURE
Aug 03 22:52:09 ip-172-31-21-132 systemd[1]: vault.service: Unit entered failed state.
Aug 03 22:52:09 ip-172-31-21-132 systemd[1]: vault.service: Failed with result 'exit-code'.
Aug 03 22:52:09 ip-172-31-21-132 systemd[1]: vault.service: Service hold-off time over, scheduling restart.
Aug 03 22:52:09 ip-172-31-21-132 systemd[1]: Stopped Vault service.
```
 
For communication with Dynamo for the Vault Backend, I use instance roles, and it looks like Vault stopped being able to communicate with Dynamo:

```
Aug  3 21:26:46 ip-172-31-21-132 dhclient[1014]: DHCPREQUEST of 172.31.21.132 on eth0 to 172.31.16.1 port 67 (xid=0x2f3edea6)
Aug  3 21:26:46 ip-172-31-21-132 dhclient[1014]: DHCPACK of 172.31.21.132 from 172.31.16.1
Aug  3 21:26:46 ip-172-31-21-132 dhclient[1014]: bound to 172.31.21.132 -- renewal in 1790 seconds.
Aug  3 21:47:42 ip-172-31-21-132 vault[1372]: 2018-08-03T21:47:42.121Z [ERROR] core: failed to acquire lock: error="NoCredentialProviders: no valid providers in chain. Deprecated.
Aug  3 21:47:42 ip-172-31-21-132 vault[1372]: #011For verbose messaging see aws.Config.CredentialsChainVerboseErrors"
Aug  3 21:47:45 ip-172-31-21-132 vault[1372]: 2018-08-03T21:47:45.105Z [ERROR] core: key rotation periodic upgrade check failed: error="NoCredentialProviders: no valid providers in chain. Deprecated.
Aug  3 21:47:45 ip-172-31-21-132 vault[1372]: #011For verbose messaging see aws.Config.CredentialsChainVerboseErrors"
// message repeates hundreds of times
``` 

The big issue here is that when Vault couldn't communicate with the backend it didn't die or go into standby mode. The other Vault instance was still capable of communicating with Dynamo, but I don't think it was promoted to leader. This prevented any other service from communicating with Vault, even though a standby was still available.

**To Reproduce**
Steps to reproduce the behavior: (this is what I *think* happened)
1. Start Vault in HA mode with the DynamoDB backend with separate AWS credentials.
2. Revoke the AWS credentials on the Vault Leader
3. The leader refuses to service requests, but does not step down either.

**Expected behavior**
The leader should die or step down.

**Environment:**
* Vault Server Version (retrieve with `vault status`): 0.10.3
* Vault CLI Version (retrieve with `vault version`): 0.10.3
* Server Operating System/Architecture: ubuntu-xenial-16.04

Vault server configuration file(s):
```hcl
storage "dynamodb" {
  ha_enabled = "true"
  region = "us-east-2"
  table = "vault-dynamodb-backend"
}

listener "tcp" {
  address     = "0.0.0.0:8200"
}

telemetry {
  statsd_address = "telegraf.corp:8125"
}

ui = true
```

**Additional context**
There may be a second bug where Vault doesn't refresh its credentials with Instance Profiles properly, but I'm not sure if thats a Vault issue, or if I just haven't configured my Instance Profile correctly.

EDIT: It looks like I've hit some AWS issue where I can on longer communicate with the metadata server on that node:

```sh
ubuntu@ip-172-31-21-132:~$ curl http://169.254.169.254/latest/meta-data/hostname;
curl: (7) Failed to connect to 169.254.169.254 port 80: Connection refused
```

So a node got knocked offline due to a cloud issue, but didn't properly step down.

**Comment by spangenberg** [comment]
I just tested the expected behavior with Vault v1.3.2, and it seems to works accurately.

The standby node became active shortly after I revoked the credentials of the before active node.

But it's important to point out that the node with the now revoked credentials responds to every request with an HTTP 500 error:
```
UnrecognizedClientException: The security token included in the request is invalid.
        status code: 400, request id: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

It can't go into standby since this is a typical Split-brain scenario.

I hope this helps you further. Feel free to respond in case there is more to it.
